gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -447,16 +572,30 @@ private ManagedChannel createSingleChannel() throws IOException {
       builder.keepAliveTime(DIRECT_PATH_KEEP_ALIVE_TIME_SECONDS, TimeUnit.SECONDS);
       builder.keepAliveTimeout(DIRECT_PATH_KEEP_ALIVE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
     } else {
+      // Try and create credentials via DCA. See https://google.aip.dev/auth/4114.
       ChannelCredentials channelCredentials;
       try {
         channelCredentials = createMtlsChannelCredentials();
       } catch (GeneralSecurityException e) {
         throw new IOException(e);
       }
       if (channelCredentials != null) {
+        // Create the channel using channel credentials created via DCA.
         builder = Grpc.newChannelBuilder(endpoint, channelCredentials);
       } else {
-        builder = ManagedChannelBuilder.forAddress(serviceAddress, port);
+        // Could not create channel credentials via DCA. In accordance with
+        // https://google.aip.dev/auth/4115, if credentials not available through
+        // DCA, try mTLS with credentials held by the S2A (Secure Session Agent).
+        if (endpointContext.useS2A()) {
+          channelCredentials = createS2ASecuredChannelCredentials();
+        }
+        if (channelCredentials != null) {
+          // Create the channel using S2A-secured channel credentials.
+          builder = Grpc.newChannelBuilder(endpointContext.mtlsEndpoint(), channelCredentials);
```

Comment:
We may still need to pass `useS2A()` to `InstantiatingGrpcChannelProvider`, but that is better than passing `EndpointContext` or passing both `useS2A()` and `mtlsEndpoint`, `EndpointContext.useS2A()` also does not have to be public now.

Reply:
Is my following understanding correct?
> Yes. If `endpointContext.useS2A() == true`, than it would be correct to have the `resolvedEndpoint()` just return the `mtlsEndpoint`

We may still need to pass useS2A() to InstantiatingGrpcChannelProvider, but that is better than passing EndpointContext or passing both useS2A() and mtlsEndpoint
> Correct, we will need to pass `useS2A()` to the InstantiatingGrpcChannelProvider (which means that the transportChannelProvider interface will still need to be modified). If we go with this way of setting the `resolvedEndpoing` when `useS2A() == true` in `endpointContext`, you are correct that we will not need to pass in `mtlsEndpoint` or `endpointContext`

EndpointContext.useS2A() also does not have to be public now.
> Yes, we can make it package private. We can just pass the boolean to InstantiatingGrpcChannelProvider

Reply:
Let me know if this SG, and I can go ahead and implement what you have suggested. Thanks!

Reply:
Done in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/791ab2c6b570a3e9db0ca86f3998454c086d93f7